<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Logo de Skills Network">
    </a>
</p>


<h1>Extracción y Visualización de Datos de Acciones</h1>
<h2>Descripción</h2>


Extraer datos esenciales de un conjunto de datos y mostrarlos es una parte necesaria de la ciencia de datos; por lo tanto, las personas pueden tomar decisiones correctas basadas en los datos. En esta tarea, extraerás algunos datos de acciones y luego mostrarás estos datos en un gráfico.


<h2>Tabla de Contenidos</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Definir una Función que Cree un Gráfico</li>
        <li>Pregunta 1: Usar yfinance para Extraer Datos de Acciones</li>
        <li>Pregunta 2: Usar Webscraping para Extraer Datos de Ingresos de Tesla</li>
        <li>Pregunta 3: Usar yfinance para Extraer Datos de Acciones</li>
        <li>Pregunta 4: Usar Webscraping para Extraer Datos de Ingresos de GME</li>
        <li>Pregunta 5: Graficar el Gráfico de Acciones de Tesla</li>
        <li>Pregunta 6: Graficar el Gráfico de Acciones de GameStop</li>
    </ul>
<p>
    Tiempo Estimado Necesario: <strong>30 min</strong></p>
</div>

<hr>


***Nota***:- Si estás trabajando localmente usando anaconda, por favor descomenta el siguiente código y ejecútalo. Usa la versión según tu versión de Python.


In [1]:
!pip install yfinance
!pip install bs4
!pip install nbformat
!pip install --upgrade plotly

In [2]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [3]:
import plotly.io as pio
pio.renderers.default = "iframe"

En Python, puedes ignorar advertencias usando el módulo `warnings`. Puedes usar la función `filterwarnings` para filtrar o ignorar mensajes de advertencia específicos o categorías.


In [4]:
import warnings
# Ignorar todas las advertencias
warnings.filterwarnings("ignore", category=FutureWarning)

## Definir Función de Graficación


En esta sección, definimos la función `make_graph`. **No necesitas saber cómo funciona la función, solo debes preocuparte por las entradas. Toma un DataFrame con datos de acciones (el DataFrame debe contener las columnas Date y Close), un DataFrame con datos de ingresos (el DataFrame debe contener las columnas Date y Revenue), y el nombre de la acción.**


In [5]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Precio Histórico de la Acción", "Ingresos Históricos"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021-06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Precio de la Acción"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Ingresos"), row=2, col=1)
    fig.update_xaxes(title_text="Fecha", row=1, col=1)
    fig.update_xaxes(title_text="Fecha", row=2, col=1)
    fig.update_yaxes(title_text="Precio ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Ingresos ($US Millones)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()
    from IPython.display import display, HTML
    fig_html = fig.to_html()
    display(HTML(fig_html))

Usa la función `make_graph` que ya hemos definido. Tendrás que invocarla en las preguntas 5 y 6 para mostrar los gráficos y crear el panel de control. 
> **Nota: No necesitas redefinir la función para graficar gráficos en ninguna otra parte de este cuaderno; solo usa la función existente.**


## Pregunta 1: Usar yfinance para Extraer Datos de Acciones


Usando la función `Ticker` ingresa el símbolo de la acción de la que queremos extraer datos para crear un objeto ticker. La acción es Tesla y su símbolo de ticker es `TSLA`.


In [6]:
tsla = yf.Ticker("TSLA")

Usando el objeto ticker y la función `history` extrae información de la acción y guárdala en un DataFrame llamado `tesla_data`. Establece el parámetro `period` en ` "max" ` para obtener información por el máximo tiempo posible.


In [7]:
tesla_data = tsla.history(period="max")

**Restablece el índice** usando la función `reset_index(inplace=True)` en el DataFrame `tesla_data` y muestra las primeras cinco filas del DataFrame `tesla_data` usando la función `head`. Toma una captura de pantalla de los resultados y del código desde el inicio de la Pregunta 1 hasta los resultados a continuación.


In [9]:
tesla_data.reset_index(inplace=True)
tesla_data.head()

index                      Date      Open      High       Low     Close  \
0      0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1      1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2      2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3      3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4      4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0

## Pregunta 2: Usar Webscraping para Extraer Datos de Ingresos de Tesla


Usa la biblioteca `requests` para descargar la página web https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Guarda el texto de la respuesta como una variable llamada `html_data`.


In [10]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
html_data = requests.get(url).text

Analiza los datos HTML usando `beautiful_soup` con un parser, es decir, `html5lib` o `html.parser`.


In [11]:
soup = BeautifulSoup(html_data, "html.parser")

Usando `BeautifulSoup` o la función `read_html`, extrae la tabla con los `Ingresos de Tesla` y guárdala en un DataFrame llamado `tesla_revenue`. El DataFrame debe tener las columnas `Date` y `Revenue`.


<details><summary>Instrucciones paso a paso</summary>

```

Aquí están las instrucciones paso a paso:

1. Crear un DataFrame Vacío
2. Encontrar la Tabla Relevante
3. Verificar la Tabla de Ingresos Trimestrales de Tesla
4. Iterar a Través de las Filas en el Cuerpo de la Tabla
5. Extraer Datos de las Columnas
6. Añadir Datos al DataFrame

```
</details>


<details><summary>Haz clic aquí si necesitas ayuda para localizar la tabla</summary>

```
    
A continuación se muestra el código para aislar la tabla, ahora necesitarás recorrer las filas y columnas como en el laboratorio anterior
    
soup.find_all("tbody")[1]
    
Si quieres usar la función read_html, la tabla está ubicada en el índice 1

Nos estamos enfocando en los ingresos trimestrales en el laboratorio.
```

</details>


In [18]:
# Paso 1: Crear un DataFrame vacío
tesla_revenue = pd.DataFrame(columns=["Date", "Revenue"])

# Paso 2 Encontrar la tabla relevante
tables = soup.find_all("table")
print(f"Número de tablas encontradas: {len(tables)}")

# Inspeccionar cada tabla
for i, table in enumerate(tables):
    print(f"Tabla {i}:")
    print(table.prettify())

# 3: verificar que sea la de ingresos trimestrales de Tesla
# Según la pista, la tabla está en soup.find_all("tbody")[1]
table = soup.find_all("tbody")[1]

# Paso 4: Iterar a través de las filas en el cuerpo de la tabla
for row in table.find_all("tr"):  # Recorre todas las filas <tr>
    cols = row.find_all("td")  # Extrae todas las celdas <td> de la fila
    
    # Paso 5: Extraer datos de las columnas
    # Asumimos que la primera columna es Date y la segunda es Revenue
    if len(cols) >= 2:  # Verifica que haya al menos 2 columnas
        date = cols[0].text.strip()  # Fecha en la primera celda
        revenue = cols[1].text.strip()  # Ingresos en la segunda celda
        
        # Paso 6: Añadir datos al DataFrame
        tesla_revenue = pd.concat([tesla_revenue, pd.DataFrame({"Date": [date], "Revenue": [revenue]})], ignore_index=True)

# Verificar el resultado
tesla_revenue.head()

Número de tablas encontradas: 6
Tabla 0:
<table class="historical_data_table table">
 <thead>
  <tr>
   <th colspan="2" style="text-align:center">
    Tesla Annual Revenue
    <br/>
    <span style="font-size:14px;">
     (Millions of US $)
    </span>
   </th>
  </tr>
 </thead>
 <tbody>
  <tr>
   <td style="text-align:center">
    2021
   </td>
   <td style="text-align:center">
    $53,823
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2020
   </td>
   <td style="text-align:center">
    $31,536
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2019
   </td>
   <td style="text-align:center">
    $24,578
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2018
   </td>
   <td style="text-align:center">
    $21,461
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2017
   </td>
   <td style="text-align:center">
    $11,759
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2016
   </td>
   <td style="text-align:center">
    $7,0

Date  Revenue
0  2022-09-30  $21,454
1  2022-06-30  $16,934
2  2022-03-31  $18,756
3  2021-12-31  $17,719
4  2021-09-30  $13,757

Ejecuta la siguiente línea para eliminar la coma y el signo de dólar de la columna `Revenue`. 


In [19]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"",regex=True)

Ejecuta las siguientes líneas para eliminar valores nulos o cadenas vacías en la columna `Revenue`.


In [20]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]

Muestra las últimas 5 filas del DataFrame `tesla_revenue` usando la función `tail`. Toma una captura de pantalla de los resultados.


In [21]:
tesla_revenue.tail()

Date Revenue
48  2010-09-30      31
49  2010-06-30      28
50  2010-03-31      21
52  2009-09-30      46
53  2009-06-30      27

## Pregunta 3: Usar yfinance para Extraer Datos de Acciones


Usando la función `Ticker` ingresa el símbolo de la acción de la que queremos extraer datos para crear un objeto ticker. La acción es GameStop y su símbolo de ticker es `GME`.


In [26]:
import yfinance as yf
gme = yf.Ticker("GME")

Usando el objeto ticker y la función `history` extrae información de la acción y guárdala en un DataFrame llamado `gme_data`. Establece el parámetro `period` en ` "max" ` para obtener información por el máximo tiempo posible.


In [28]:
gme_data=gme.history(period="max")

**Restablece el índice** usando la función `reset_index(inplace=True)` en el DataFrame `gme_data` y muestra las primeras cinco filas del DataFrame `gme_data` usando la función `head`. Toma una captura de pantalla de los resultados y del código desde el inicio de la Pregunta 3 hasta los resultados a continuación.


In [31]:
gme_data.reset_index(inplace=True)
gme_data.head()

Date      Open      High       Low     Close    Volume  \
0 2002-02-13 00:00:00-05:00  1.620129  1.693350  1.603296  1.691667  76216000   
1 2002-02-14 00:00:00-05:00  1.712707  1.716074  1.670626  1.683250  11021600   
2 2002-02-15 00:00:00-05:00  1.683250  1.687458  1.658001  1.674834   8389600   
3 2002-02-19 00:00:00-05:00  1.666418  1.666418  1.578047  1.607504   7410400   
4 2002-02-20 00:00:00-05:00  1.615920  1.662210  1.603296  1.662210   6892800   

   Dividends  Stock Splits  
0        0.0           0.0  
1        0.0           0.0  
2        0.0           0.0  
3        0.0           0.0  
4        0.0           0.0

## Pregunta 4: Usar Webscraping para Extraer Datos de Ingresos de GME


Usa la biblioteca `requests` para descargar la página web https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Guarda el texto de la respuesta como una variable llamada `html_data_2`.


In [32]:
url = "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html""(Optional) HTML for Web Scraping.txt"
html_data2 = requests.get(url).text

Analiza los datos HTML usando `beautiful_soup` con un parser, es decir, `html5lib` o `html.parser`.


In [ ]:
soup2 = BeautifulSoup(html_data2, "html.parser")

C:\Users\Usuario\AppData\Local\Temp\ipykernel_20396\1752612562.py:5: XMLParsedAsHTMLWarning:

It looks like you're using an HTML parser to parse an XML document.

Assuming this really is an XML document, what you're doing might work, but you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the Python package 'lxml' installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.

If you want or need to use an HTML parser on this document, you can make this warning go away by filtering it. To do that, run this code before calling the BeautifulSoup constructor:

    from bs4 import XMLParsedAsHTMLWarning
    import warnings

    warnings.filterwarnings("ignore", category=XMLParsedAsHTMLWarning)




Usando `BeautifulSoup` o la función `read_html`, extrae la tabla con los `Ingresos de GameStop` y guárdala en un DataFrame llamado `gme_revenue`. El DataFrame debe tener las columnas `Date` y `Revenue`. Asegúrate de que la coma y el signo de dólar se eliminen de la columna `Revenue`.


> **Nota: Usa el método similar al que usaste en la pregunta 2.**  


<details><summary>Haz clic aquí si necesitas ayuda para localizar la tabla</summary>

```
    
A continuación se muestra el código para aislar la tabla, ahora necesitarás recorrer las filas y columnas como en el laboratorio anterior
    
soup.find_all("tbody")[1]
    
Si quieres usar la función read_html, la tabla está ubicada en el índice 1


```

</details>


Muestra las últimas cinco filas del DataFrame `gme_revenue` usando la función `tail`. Toma una captura de pantalla de los resultados.


## Pregunta 5: Graficar el Gráfico de Acciones de Tesla


Usa la función `make_graph` para graficar los Datos de Acciones de Tesla, también proporciona un título para el gráfico. Nota que el gráfico solo mostrará datos hasta junio de 2021.


<details><summary>Sugerencia</summary>

```

Solo necesitas invocar la función make_graph con el parámetro requerido para imprimir los gráficos. La estructura para llamar a la función `make_graph` es `make_graph(tesla_data, tesla_revenue, 'Tesla')`.

```
    
</details>


## Pregunta 6: Graficar el Gráfico de Acciones de GameStop


Usa la función `make_graph` para graficar los Datos de Acciones de GameStop, también proporciona un título para el gráfico. La estructura para llamar a la función `make_graph` es `make_graph(gme_data, gme_revenue, 'GameStop')`. Nota que el gráfico solo mostrará datos hasta junio de 2021.


<details><summary>Sugerencia</summary>

```

Solo necesitas invocar la función make_graph con el parámetro requerido para imprimir los gráficos. La estructura para llamar a la función `make_graph` es `make_graph(gme_data, gme_revenue, 'GameStop')`

```
    
</details>


<h2>Sobre los Autores:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> tiene un PhD en Ingeniería Eléctrica, su investigación se centró en usar aprendizaje automático, procesamiento de señales y visión por computadora para determinar cómo los videos impactan la cognición humana. Joseph ha estado trabajando para IBM desde que completó su PhD.

Azim Hirjani


## Registro de Cambios

| Fecha (AAAA-MM-DD) | Versión | Modificado Por    | Descripción del Cambio        |
| ----------------- | ------- | ----------------- | ---------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla     | Cambió la URL de GameStop    |
| 2020-11-10        | 1.1     | Malika Singla     | Eliminó la parte Opcional    |
| 2020-08-27        | 1.0     | Malika Singla     | Añadió el laboratorio a GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. Todos los derechos reservados. <h3/>

<p>
